agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java

```diff
@@ -16,53 +16,80 @@
 
 package io.opentelemetry.auto.bootstrap;
 
+import io.opentelemetry.auto.bootstrap.instrumentation.api.Pair;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.ref.WeakReference;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.net.URLStreamHandler;
-import java.nio.file.NoSuchFileException;
 import java.security.Permission;
 import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import lombok.extern.slf4j.Slf4j;
 
 @Slf4j
 public class InternalJarURLHandler extends URLStreamHandler {
-  private final Map<String, JarEntry> filenameToEntry = new HashMap<>();
-  private JarFile bootstrapJarFile;
+
+  private static final WeakReference<Pair<String, JarEntry>> NULL = new WeakReference<>(null);
+
+  private final String name;
+  private final FileNotInInternalJar notFound;
```

Comment:
If I understand correctly, all these changes to this class are performance optimisations, no? Is there any benchmark or, even better, a real use-case, which demonstrate the benefit of this?